### PR TITLE
Add Max Consecutive Ones to CTCI 10092020

### PR DESCRIPTION
**Pull request by iambillzhao**

### CODE REVIEW

File: 2020-03-Autumn/0_sliding_window/README.md [hunk 1]
```diff
@@ -94,21 +105,62 @@ Go to [Solution](#s3)   [Top](#top)
...
+    return max;
+}
+```
 
 #### Optimal Solution
 
-TODO :bug:
+Instead of counting the maximum numbers of 1s in a group, we can also define 
+where the group is for faster calculation. In the following solution, the program 
+uses two indecies to keep track of the start and the end of a group of 1s. This is
```
Comment (hollowsunsets) [inline 2020-03-Autumn/0_sliding_window/README.md:141]:
```suggestion
uses two indices to keep track of the start and the end of a group of 1s. This is 
```

File: 2020-03-Autumn/0_sliding_window/README.md [hunk 2]
```diff
@@ -94,21 +105,62 @@ Go to [Solution](#s3)   [Top](#top)
...
 
 #### Optimal Solution
 
-TODO :bug:
+Instead of counting the maximum numbers of 1s in a group, we can also define 
+where the group is for faster calculation. In the following solution, the program 
+uses two indecies to keep track of the start and the end of a group of 1s. This is 
+called the "Sliding window" technique. You can move this window around while going 
+through the binary array, checking its value and update the maximum variable.
+
```
Comment (hollowsunsets) [inline 2020-03-Autumn/0_sliding_window/README.md:144]: Maybe mention the time complexity? 
Specifically how it's gone from `O(N^2)` ---> `O(N)`

File: 2020-03-Autumn/0_sliding_window/README.md [hunk 3]
```diff
@@ -94,21 +105,62 @@ Go to [Solution](#s3)   [Top](#top)
...
+#### Brute Force Solution
+
+At the first glance of this question, we can see that the program wants to 
+count the most 1s that are grouped together. As we go through the binary 
+array, we can count along the way: If it is a 0, we would stop counting 
+and reset the counter to 0. Otherwise, we would increment the counter by 1. 
+The maximum number of the counter will always be stored in a variable called 
+`max`.
+
```
Comment (hollowsunsets) [inline 2020-03-Autumn/0_sliding_window/README.md:120]: This isn't quite the brute force solution. I'd actually say it's another way of solving this problem optimally in `O(N)` time without a sliding window.

If you can, I'd try to add the `O(N^2)` brute force solution - something like this:

```
max_count = 0
n = len(arr)
for i in range(n):
 if i == 1:
    count = 0
    for j in range(i, n):
        if arr[j] == 1:
           count += 1
        else:
            max_count = max(count, max_count)
            break
return max_count
```

File: 2020-03-Autumn/0_sliding_window/QuestionOne/driver.cpp [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            max = std::max(count, max);
+        } else {
+            count = 0;
+        }
+    }
+    return max;
+}
+
+int main() {
+    std::vector<int> four (11,0);
```
Comment (hollowsunsets) [inline 2020-03-Autumn/0_sliding_window/QuestionOne/driver.cpp:24]: Thanks for adding a test :D

**Review by hollowsunsets — CHANGES_REQUESTED** [review]
Overall, looks good.  I'd recommend adding the O(N^2) brute force solution to make it clear what the benefit of using the sliding window technique is.

File: 2020-03-Autumn/0_sliding_window/README.md (hunk 2 above)
Comment (iambillzhao) [reply 2020-03-Autumn/0_sliding_window/README.md:173]: Changes addressed in commit 8a11068.

File: 2020-03-Autumn/0_sliding_window/README.md (hunk 3 above)
Comment (iambillzhao) [reply 2020-03-Autumn/0_sliding_window/README.md:148]: Changes addressed in commit 8a11068.

File: 2020-03-Autumn/0_sliding_window/QuestionOne/driver.cpp (hunk 4 above)
Comment (iambillzhao) [reply 2020-03-Autumn/0_sliding_window/QuestionOne/driver.cpp:24]: :) Go CTCI!

**Review by hollowsunsets — APPROVED** [review]

**Pull request merged**